image/generate.go

```diff
@@ -0,0 +1,362 @@
+}
+
+func createRefFile(root, ref string, mft descriptor) error {
+	refpath := filepath.Join(root, "refs", ref)
+	f, err := os.Create(refpath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	refStr = strings.Replace(refStr, "<manifest_digest>", mft.Digest, -1)
+	refStr = strings.Replace(refStr, "<manifest_size>", strconv.FormatInt(mft.Size, 10), -1)
+	_, err = io.Copy(f, bytes.NewBuffer([]byte(refStr)))
+	return err
+}
+
+func createManifestFile(root, str string) (descriptor, error) {
+	name := filepath.Join(root, "blobs", "sha256", "test-manifest")
+	f, err := os.Create(name)
+	if err != nil {
+		return descriptor{}, err
+	}
+	defer f.Close()
+
+	_, err = io.Copy(f, bytes.NewBuffer([]byte(str)))
+	if err != nil {
+		return descriptor{}, err
+	}
+
+	return createHashedBlob(name)
+}
+
+func createConfigFile(root, config string) (descriptor, error) {
+	name := filepath.Join(root, "blobs", "sha256", "test-config")
+	f, err := os.Create(name)
+	if err != nil {
+		return descriptor{}, err
+	}
+	defer f.Close()
+
+	_, err = io.Copy(f, bytes.NewBuffer([]byte(config)))
+	if err != nil {
+		return descriptor{}, err
+	}
+
+	return createHashedBlob(name)
+}
+
+func createImageLayerFile(root string, list []tarContent) (descriptor, error) {
+	name := filepath.Join(root, "blobs", "sha256", "test-layer")
+	err := createTarBlob(name, list)
+	if err != nil {
+		return descriptor{}, err
+	}
+
+	desc, err := createHashedBlob(name)
+	if err != nil {
+		return descriptor{}, err
+	}
+
+	desc.MediaType = v1.MediaTypeImageLayer
+	return desc, nil
+}
+
+func createTarBlob(name string, list []tarContent) error {
+	file, err := os.Create(name)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	gzipWriter := gzip.NewWriter(file)
+	defer gzipWriter.Close()
+	tarWriter := tar.NewWriter(gzipWriter)
+	defer tarWriter.Close()
+
+	for _, content := range list {
+		if err = tarWriter.WriteHeader(content.header); err != nil {
+			return err
+		}
+		if _, err = io.Copy(tarWriter, bytes.NewReader(content.b)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func createHashedBlob(name string) (descriptor, error) {
+	desc, err := newDescriptor(name)
+	if err != nil {
+		return descriptor{}, err
+	}
+
+	// Rename the file to hashed-digest name.
+	err = os.Rename(name, filepath.Join(filepath.Dir(name), desc.Digest))
+	if err != nil {
+		return descriptor{}, err
+	}
+
+	//Normalize the hashed digest.
+	desc.Digest = "sha256:" + desc.Digest
+
+	return desc, nil
+}
+
+func newDescriptor(name string) (descriptor, error) {
+	file, err := os.Open(name)
+	if err != nil {
+		return descriptor{}, err
+	}
+	defer file.Close()
+
+	// generate sha256 hash
+	hash := sha256.New()
+	size, err := io.Copy(hash, file)
+	if err != nil {
+		return descriptor{}, err
+	}
+
+	return descriptor{
+		Digest: fmt.Sprintf("%x", hash.Sum(nil)),
+		Size:   size,
+	}, nil
+}
+
+//Compress use gzip to compress tar.gz
+func Compress(src string) error {
+	d, err := os.Create(src + ".tar")
```

Comment:
A function that compresses an on-disk directory into an on-disk file (I don't actually see any code in here about compressing with gzip) doesn't seem like the friendliest public API.  And you don't seem to be using this function anywhere.  So I suspect you should drop `Compress` and `compress` from this PR.